bus/accounts.go

```diff
@@ -147,35 +147,43 @@ func (a *accounts) AddAmount(id rhpv3.Account, hk types.PublicKey, amt *big.Int)
 }
 
 // SetBalance sets the balance of a given account to the provided amount. If the
-// account doesn't exist, it is created.
-// If an account hasn't been saved successfully upon the last shutdown, no drift
-// will be added upon the first call to SetBalance.
+// account doesn't exist, it is created. If an account hasn't been saved
+// successfully upon the last shutdown, no drift will be added upon the first
+// call to SetBalance.
 func (a *accounts) SetBalance(id rhpv3.Account, hk types.PublicKey, balance *big.Int) {
 	acc := a.account(id, hk)
 
-	// Update balance and drift.
 	acc.mu.Lock()
-	delta := new(big.Int).Sub(balance, acc.Balance)
-	balanceBefore := acc.Balance.String()
-	driftBefore := acc.Drift.String()
-	if acc.CleanShutdown {
-		acc.Drift = acc.Drift.Add(acc.Drift, delta)
-	}
+	defer acc.mu.Unlock()
+
+	// save previous values
+	prevBalance := new(big.Int).Set(acc.Balance)
+	prevDrift := new(big.Int).Set(acc.Drift)
+
+	// update balance
 	acc.Balance.Set(balance)
+
+	// update drift
+	drift := new(big.Int).Sub(balance, prevBalance)
+	if acc.CleanShutdown && prevBalance.Cmp(big.NewInt(0)) != 0 {
```

Comment:
If we sync an account every time we load it into memory and the shutdown was clean, we shouldn't need to check the previous balance for 0 here. Because a new account starts with `CleanShutdown = false`. So we should in theory not accumulate drift for an account we just created.

Reply:
That is true but I figured it's a good condition to have on the drift field regardless? Drift should only occur on a clean account with a non-zero balance because otherwise the drift is in our favour anyway. I don't mind changing it I just liked the documentation-aspect of having that condition guarding the `drift` field very explicitly.

Reply:
It should indeed but it's just a little more accurate. Because drift goes both ways.